utils/create_circleci_config.py

```diff
@@ -0,0 +1,371 @@
+        if self.additional_env is None:
+            self.additional_env = {}
+        if self.cache_name is None:
+            self.cache_name = self.name
+        if self.docker_image is None:
+            # Let's avoid changing the default list and make a copy.
+            self.docker_image = copy.deepcopy(DEFAULT_DOCKER_IMAGE)
+        if self.install_steps is None:
+            self.install_steps = []
+        if self.pytest_options is None:
+            self.pytest_options = {}
+        if isinstance(self.tests_to_run, str):
+            self.tests_to_run = [self.tests_to_run]
+
+    def to_dict(self):
+        job = {
+            "working_directory": self.working_directory,
+            "docker": self.docker_image,
+            "environment": {**COMMON_ENV_VARIABLES, **self.additional_env},
+        }
+        if self.resource_class is not None:
+            job["resource_class"] = self.resource_class
+        if self.parallelism is not None:
+            job["parallelism"] = self.parallelism
+        steps = [
+            "checkout",
+            {"attach_workspace": {"at": "~/transformers/test_preparation"}},
+            {
+                "restore_cache": {
+                    "keys": [
+                        f"v{self.cache_version}-{self.cache_name}-" + '{{ checksum "setup.py" }}',
+                        f"v{self.cache_version}-{self.cache_name}-",
+                    ]
+                }
+            },
+        ]
+        steps.extend([{"run": l} for l in self.install_steps])
+        steps.append(
+            {
+                "save_cache": {
+                    "key": f"v{self.cache_version}-{self.cache_name}-" + '{{ checksum "setup.py" }}',
+                    "paths": ["~/.cache/pip"],
+                }
+            }
+        )
+
+        all_options = {**COMMON_PYTEST_OPTIONS, **self.pytest_options}
+        pytest_flags = [f"--{key}={value}" if value is not None else f"-{key}" for key, value in all_options.items()]
+        pytest_flags.append(
+            f"--make-reports={self.name}" if "examples" in self.name else f"--make-reports=tests_{self.name}"
+        )
+        test_command = f"python -m pytest -n {self.pytest_num_workers} " + " ".join(pytest_flags)
+        if self.tests_to_run is None:
+            test_command += " << pipeline.parameters.tests_to_run >>"
+        else:
+            test_command += " " + " ".join(self.tests_to_run)
+        if self.marker is not None:
+            test_command += f" -m {self.marker}"
+        test_command += " | tee tests_output.txt"
+        steps.append({"run": {"name": "Run tests", "command": test_command}})
+        steps.append({"store_artifacts": {"path": "~/transformers/tests_output.txt"}})
+        steps.append({"store_artifacts": {"path": "~/transformers/reports"}})
+        job["steps"] = steps
+        return job
+
+    @property
+    def job_name(self):
+        return self.name if "examples" in self.name else f"tests_{self.name}"
+
+
+# JOBS
```

Comment:
Could this exist in a script under `.circleci/`? That's the place where we would logically go take a look to see where the tests are generated. I wouldn't take a look in `utils`, so I think it's a good place to store the generation.